### DISCUSSION TITLE
final TypedDicts with total=False does not allow clear/popitem

### DISCUSSION THREAD

**Discussion by owocado** [microsoft/pyright#9369] [General]
Hello,

![image](https://github.com/user-attachments/assets/778bc22f-cb91-4705-a044-16139b58f3d0)

pyright version: 1.1.387
python version: 3.12.7
vscode version: 1.95.0

I have searched for this issue on internet and I have found https://github.com/microsoft/pyright/issues/4794 which suggests that this got fixed back in 2023 in pyright but I can reproduce this now so can someone please guide me here on what changed in pyright since then to cause this?
If I missed something else useful of context on this then please do let me know.

I have also checked the two issues mentioned in the above issue,
as well as I have also read the rationale mentioned on:
- https://typing.readthedocs.io/en/latest/spec/typeddict.html#assignability
- https://typing.readthedocs.io/en/latest/spec/typeddict.html#supported-and-unsupported-operations

as to why `TypedDict` generally do not allow `clear()` methods to safeguard us for potential aspect of destructive operations,

however

my use cases of `TypedDict` in my small projects are casual enough so I do not have to account for such rationales so is there any way I can `.clear()` in peace or is this a valid bug report w.r.t. pyright?

Thanks for reading and have a nice day!

**Comment by erictraut** [comment]
Could you please provide your code sample in text form either here or in the [pyright playground](https://pyright-play.net/) so I don't need to retype it from a screen shot?

**Reply by erictraut** [reply]
The `clear` and `popitem` methods are not type safe to call on a TypedDict unless it is closed (i.e. not allowed to have any additional items beyond those that are defined). There's currently no way to specify that a TypedDict is closed, but draft [PEP 728](https://peps.python.org/pep-0728/) adds this capability. Pyright supports this as an experimental feature, but I don't recommend using it in production code until after the PEP has been accepted.

In the meantime, if you want to ignore this unsafe behavior in your code, you can use a `# type: ignore` or `# pyright: ignore[reportAttributeAccessIssue]` comment to suppress the type checking error in your code.

**Reply by owocado** [reply]
yeah I agree with that rationale why that is not allowed with TypedDict however this does not address one of my original question about why pyright is causing this error now because the issue I linked in my original post states that support for clear/popitem was added in 2023 in pyright so can I ask again what has changed in pyright since then to revert that support?

I am also aware that I can momentarily fix this pyright error with your suggestion of `# type: ignore` but I want to avoid doing that as I will have to apply it in so many places (long story, my project works with and generate huge dictionaries at various places and they're typed and I need to free the allocated resources at the end of function logic to reduce any memory leaks).
same dichotomy with the use of `reportAttributeAccessIssue` as well unless I add it in my pyproject file to get rid of the error project wide but that would prevent pyright reporting other valid cases of attribute access issues.
Is there no other workaround for this temporarily so it let me `.clear()` in peace?

**Reply by erictraut** [reply]
> what has changed in pyright since then to revert that support?

Since then, the correct (type safe) behavior was clarified in the typing spec and among type checker authors. Pyright is compliant with the typing spec.

---

> Is there no other workaround for this temporarily so it let me .clear() in peace?

You could write a helper function that calls the `clear` method and then apply the `# type: ignore` only in that one location. Or you can disable the `reportAttributeAccessIssue` for your project.

**Reply by owocado** [reply]
ok thanks, eric. I'll go with the helper function route. have a nice day!

**Comment by owocado** [comment]
hello eric,

yeah sure

```py
import typing

@typing.final
class Foo(typing.TypedDict, total=False):
    a: int
    b: int
    c: list[str]


a: Foo = {'a': 1, 'b': 2, 'c': ['a', 'b']}
a.clear()
a.popitem()
```

btw thank you for replying this promptly :D
